### ISSUE TITLE
Ability to use and override default profile settings in a custom profile

### ISSUE THREAD

**Issue by sandy081** [microsoft/vscode#188612]
Ability to use and override default profile settings in a custom profile

**Comment by tanishqmanuja** [comment]
Use case: https://github.com/microsoft/vscode/issues/156144#issuecomment-1642405219

**Comment by tanishqmanuja** [comment]
I have recently updated my approach towards maintaining my vscode settings.
Purpose of sharing it here is:
- Feedback for improving/ Discussing any alternative method
- Reporting the pain points in this approach to vscode devs

![vsc-profiles](https://github.com/microsoft/vscode/assets/40914272/9cad5295-f880-4585-9a6c-457794ea7eb9)

### Pain Points
These are marked with red arrow

- Have to copy all defaults manually
- No way of enabling extensions for workspace via code/json. Code can be copy-pasted/automated but not GUI actions.

**Comment by cdce8p** [comment]
> No way of enabling extensions for workspace via code/json. Code can be copy-pasted/automated but not GUI actions.

Would love to do that as well. I frequently need to enable the same extensions for new (temporary) workspaces.
In theory it would be possible to edit the `state.vscdb` database files and overwrite `extensionsIdentifiers/enabled` but that's something which should be simpler to accomplish.

**Comment by jcrben** [comment]
The use case I envision is for team sharing of settings. I want to create base default settings for all my team members which they can easily override if they feel so inclined. Then I want them to be able to use a profile as well on top of this foundation.

**Comment by thernstig** [comment]
I believe the solution should be akin ESLint extends. The current way of using `workbench.settings.applyToAllProfiles` is not sufficient to capture everything. Here is an example of this shortcoming:

Imagine I have this in the applications setting (**default profile**):

```jsonc
{
  "workbench.settings.applyToAllProfiles": [
    "editor.codeActionsOnSave"
  ],
  "editor.codeActionsOnSave": {
    "source.organizeImports": "never"
  }
}
```

Imagine I have this in a **web profile** setting.json:

```jsonc
{
  "editor.codeActionsOnSave": {
    "source.organizeImports": "always",
    "source.fixAll.eslint": "always",
    "source.fixAll.markdownlint": "always",
    "source.fixAll.stylelint": "always"
  }
}
```

There is no way for me to override the individual items in the object. Setting values can be of various types. Such merging is well handled by normal settings.json merging between user, remote, workspace settings etc, see https://code.visualstudio.com/docs/getstarted/settings#_settings-precedence.

Here is an idea of a better solution:

**Default profile**

```jsonc
  "editor.codeActionsOnSave": {
    "source.organizeImports": "never"
  }
}
```

**Default profile**

```jsonc
{
  extends: ["defaultProfile"],
  "editor.codeActionsOnSave": {
    "source.organizeImports": "always",
    "source.fixAll.eslint": "always",
    "source.fixAll.markdownlint": "always",
    "source.fixAll.stylelint": "always"
  }
}
```

This has several benefits:
* You can nest multiple "extends" where precedence is from right-to-left (i.e. last item has highest precedence). This enables a hierarchy of profiles e.g. `default -> web -> web with Golang backend/web with Python backend`.
* You easily see in the lower-layer profiles what they extend, instead of having to go to the default profile to see what gets into lower profiles.
* https://code.visualstudio.com/docs/getstarted/settings#_settings-precedence can be respected with type merging, meaning objects merge better.

**Comment by thernstig** [comment]
Think more about https://github.com/microsoft/vscode/issues/188612#issuecomment-1776797874, it might even be more beneficial to create this `extends` functionality directly built into the Profiles. Rationale being that then also Extensions, Keyboard Shortcuts etc. could be merged in a nicer way.

So let's say you have `Default -> Python Development -> Python Development Ruff` `Default -> Python Development -> Python Development Flake8`, you could share the `Python Development` and get the Python extension in both, but different linter extensions `Ruff` vs. `Flake8`. Just as a dumb example. So not only settings would be merged, but also the others. So Extensions would be a be a union of the all extended Profiles.

**Comment by UPIMMUNITY** [comment]
I wonder with current solution, when create new profile from default with settings, does it simply copy settings or references them from default?

**Comment by brettcannon** [comment]
> I wonder with current solution, when create new profile from default with settings, does it simply copy settings or references them from default?

I believe it references them.

**Comment by soimon** [comment]
I wholeheartedly agree with @thernstig on https://github.com/microsoft/vscode/issues/188612#issuecomment-1776797874: this is the way I would expect the system to work,  given my experience with various configuration files that follow the same logic.

I would like to propose the idea of multiple inheritance (with the settings being merged in order, i.e., the latter one overwriting the earlier one). This would allow you to truly separate different contexts and environments.

You could have profiles for each language and context, stacking them just like you stack your tech stack:
```ts
{
  extends: ["Svelte Profile", "Python Profile"]
  // Svelte frontend with Python backend specific settings here
}
```
```mermaid
graph TD;
    golang(Golang)
    svelte(Svelte)
    python(Python)
    edu("Presentation (larger fonts, etc.)")
    svelte & golang-->c1[Svelte with Golang backend]
    svelte & python-->c2[Svelte with Python backend]
```

In this approach, a profile becomes an expression of extensions and settings you'd like to have available when working in a specific context or using a particular technology. It's like knowing you're heading to the gym and doing groceries afterward: you simply grab your sports bag and wallet, and the union of these allows you to do all your activities without having to think too much about what's inside either of them.

To avoid having to set up a jungle of profile combinations, it makes sense to specify this union in your workspace file, as the combinationary aspect is typically specific to the workspace.

While this idea might extend beyond the initial scope of what this system was intended for, I think it offers the most flexible and (for a lot of developers) familiar approach.

**Comment by thernstig** [comment]
@soimon yes, multiple inheritance is captured in https://github.com/microsoft/vscode/issues/188612#issuecomment-1776797874 since the example uses an array for extending.

However, I think I make a fair point in https://github.com/microsoft/vscode/issues/188612#issuecomment-1781273772 that this should _possibly_ not go into the settings.json but rather Profiles directly.

**Comment by tanishqmanuja** [comment]
Since, this feature is still not supported. I am using the extension enable/disable method, can anyone tell me how to suppress these warning when the extensions are disabled.

![image](https://github.com/microsoft/vscode/assets/40914272/7bbab288-5b24-41a3-839f-99dcca17c1a8)

**Comment by tanishqmanuja** [comment]
@jmevel, Actually I dont want to switch profile at all here ( bc state of profiles is currently a mess ofc ). What I am doing is making my default profile installing all essential extensions and then disabling them as soon as they are installed. I only enable the required extension for the workspace, for example rust analyzer for rust workspace.

**Comment by thernstig** [comment]
> @jmevel, Actually I dont want to switch profile at all here ( bc state of profiles is currently a mess ofc ). What I am doing is making my default profile installing all essential extensions and then disabling them as soon as they are installed. I only enable the required extension for the workspace, for example rust analyzer for rust workspace.

The current solution is not in the best shape, as key aspects like overriding and sync between machines seem to not behave in a great way, but let's stay positive it changes in the future with our feedback :)

I made a smaller summary here of some problems: https://github.com/microsoft/vscode/issues/196718#issuecomment-1919029509

**Comment by jmevel** [comment]
> @jmevel, Actually I dont want to switch profile at all here ( bc state of profiles is currently a mess ofc ). What I am doing is making my default profile installing all essential extensions and then disabling them as soon as they are installed. I only enable the required extension for the workspace, for example rust analyzer for rust workspace.

Oh I see, that's a smart workaround, I didn't even think of it 😅

**Comment by mschwandt98** [comment]
Hi! 😊 What is the status of this feature? Has a decision already been made? If not, what exactly is missing for a decision? Many of my colleagues and I would be very happy if this feature would be implemented in VSCode.

**Comment by PierreLouisPAUGAM** [comment]
Hi @sandy081 :)
I wonder if this feature is still being worked on or if it is currently stalled.
Would you mind giving us an update ?

**Comment by gjsjohnmurray** [comment]
@PierreLouisPAUGAM FYI the team is focused on [endgame](https://github.com/microsoft/vscode/issues/217282) this week and next week.

**Comment by sandy081** [comment]
This is not in our plan yet.

**Comment by dallenbaldwin** [comment]
I found myself here after looking for a way for my profiles to inherit from a default profile. I hate having to constantly remember that some new extension I found in one profile now has to be installed in all my other profiles, and then I have to switch profiles to keep it up to date.

A pretty common use case, I would argue, is markdown. Most people have to write markdown regardless of which language or development environment they work with. There are a lot of really good markdown extensions out there for mermaid, automatic link tables of contents, printing as PDF, etc.

Another would be GitLens. It adds many features to the Source Control panel that are applicable regardless of language, dev tools, etc.

It would make sense to have a base profile hold GitLens and all the markdown extensions I like, and then, as others have mentioned, add/remove based on the framework, development environment, language, etc.

**Comment by gjsjohnmurray** [comment]
@dallenbaldwin are you aware that extensions can optionally be installed for all profiles?

https://code.visualstudio.com/docs/editor/profiles#_apply-an-extension-to-all-profiles

**Comment by dallenbaldwin** [comment]
Ah! I've never noticed that before! Thank you 😁

Does it also apply that extension to new profiles or is it only another way to copy the extension to all current profiles?

**Comment by gjsjohnmurray** [comment]
> Does it also apply that extension to new profiles

This should be easy enough to check. Also relevant is #193446 which is milestoned for the current iteration.

**Comment by gjsjohnmurray** [comment]
> Does it also apply that extension to new profiles

This should be easy enough to check. Also relevant is #193446 which is milestoned for the current iteration.

**Comment by CTNOriginals** [comment]
I have been looking for a feature like this for a while now. The way I would apply this feature is that I would create some sort of "base" profile that all profiles will inherit from so it can then serve as a sort of essentials bundle that i would apply to all profiles no matter what the profile is created and used for. This would make it a lot easier if i would find a new extension that has to be set up in a specific way so I don't have to copy paste those settings across all profiles. This would also apply to keybinds and other profile related settings.

**Comment by sanmai-NL** [comment]
> I found myself here after looking for a way for my profiles to inherit from a default profile. I hate having to constantly remember that some new extension I found in one profile now has to be installed in all my other profiles, and then I have to switch profiles to keep it up to date.
> 
> A pretty common use case, I would argue, is markdown. Most people have to write markdown regardless of which language or development environment they work with. There are a lot of really good markdown extensions out there for mermaid, automatic link tables of contents, printing as PDF, etc.
> 
> Another would be GitLens. It adds many features to the Source Control panel that are applicable regardless of language, dev tools, etc.
> 
> It would make sense to have a base profile hold GitLens and all the markdown extensions I like, and then, as others have mentioned, add/remove based on the framework, development environment, language, etc.

@dallenbaldwin 
See:

1. https://github.com/microsoft/vscode/issues/234258
1. https://github.com/microsoft/vscode/issues/40239

**Comment by alefragnani** [comment]
Hi,

First of all, congratulations for the User Profiles feature. It was a game changer!

About it, the main reason why I use profiles is **to work with different programming languages**. 

Basically, _I inherit 99% of the settings_ from my Default profiles, simply because I want to inherit the exact same VS Code experience whichever stack I'm using. The other profiles (C#, Delphi, Java, etc) just adds the settings and snippets exclusive to the extensions that are installed only on those profiles, with very few exceptions, like `workbench.colorTheme`, just for fun. 

So, in the end, all settings and keyboard shortcuts are be inherited from the Default profile, by default.

Yes, I know you asked NOT no tell _how you want to use_, but I can't hold myself 😆 . If you allows something like `"workbench.settings.applyToAllProfiles": [ "*" ]`, it would be less error prone when I make changes/updates the Default profile. And would (maybe) even fix #192745, as a collateral effect.

Thank you

**Comment by idsulik** [comment]
hi!
Is there any plan to implement the feature?)

I want to use VS Code to work with different tech stack(php/js/go/swift), and I need some common settings(theme, font size, keybindings etc.) and tech stack specific settings(for js - eslint, prettier..., for go - go format, lint...), for now I have to copy paste and keep in sync the settings

**Comment by mkvlrn** [comment]
> hi! Is there any plan to implement the feature?)
> 
> I want to use VS Code to work with different tech stack(php/js/go/swift), and I need some common settings(theme, font size, keybindings etc.) and tech stack specific settings(for js - eslint, prettier..., for go - go format, lint...), for now I have to copy paste and keep in sync the settings

@idsulik 

I will risk sounding like a bitter old man, but I don't think any plans aside from _anything_ AI are currently being seriously considered for the foreseeable future.

The release notes for 1.99 were 100% AI features and it's been a trend for the past few months.

- AI
- More AI
- I heard you like AI, dawg
- Some AI to compensate for the lack of AI
- AI with a twist*

I fear the cool features, QoL changes, and [many other requests people have been clamoring for for close to a decade now](https://github.com/microsoft/vscode/issues?q=is%3Aissue%20state%3Aopen%20sort%3Areactions-%2B1-desc) will take a backseat to the slop machine.

**Comment by Yzrsah** [comment]
Please just allow me to toggle a check box on each profile to inherit settings from the default profile. That's all I want.

I have a lot of "base configuration" settings in my default profile (500 lines) and it's difficult trying to keep the settings synchronized across every profile. Whenever I make a change to my base setup, I have to go through every profile and merge the changes manually. If the sort order somehow got changed then it becomes more difficult and now it needs a diff/merge tool just to pull the actual profile-specific stuff out. Sometimes I'm too busy to mess with it so the profiles become off-synchronized and I try to remember put it on my TODO to "Fix all the config files again". I've been putting this off for months and today I'm going to fix it again and that's why I went searching for a solution and found this issue.

I just want to set it up correctly, so it's engineered correctly, so the config changes are stable, and I don't have to keep messing with it and duplicating all the settings over and over again.

**Comment by tanishqmanuja** [comment]
What @alexjthomson achieved in roughly 500 LOC is amazing, i wonder what was stopping @sandy081 doing the same 🤔

**Comment by thernstig** [comment]
When/if this happens I believe the success of ESLint's new flat config shows that VS Code should take them same approach if starting to support this, see https://eslint.org/blog/2022/08/new-config-system-part-2/. Even though https://github.com/microsoft/vscode/issues/188612#issuecomment-1776797874 is good, flat config is even better.

**Comment by airtonix** [comment]
> What [@alexjthomson](https://github.com/alexjthomson) achieved in roughly 500 LOC is amazing, i wonder what was stopping [@sandy081](https://github.com/sandy081) doing the same 🤔

investors?

**Comment by tanishqmanuja** [comment]
> investors and about five thousand other bug tickets in the queue

I only see on argument `Investors`. I cannot understand the second argument of 5000+ issues, possibly a cover up for the first one ?

Can someone help me understand this - sort by community demands, select top 10 issues.

PS:
Also this issue is broken into 3 smaller issues, so keep that in mind if combining the 👍🏻 count or emoji count.

**Comment by artemios-git** [comment]
Inheritance would be a very nice feature

**Comment by thernstig** [comment]
I gave up on profiles. It is too bugged in multiple places, and lack of inheritance makes it not-nice-to-work-with.

**Comment by sanmai-NL** [comment]
> I gave up on profiles. It is too bugged in multiple places, and lack of inheritance makes it not-nice-to-work-with.

What do you go with now?

**Comment by thernstig** [comment]
> > I gave up on profiles. It is too bugged in multiple places, and lack of inheritance makes it not-nice-to-work-with.
> 
> What do you go with now?

I am not using an alternative. I have all extensions installed I ever want in any repo, be that Python, Rust, JavaScript, Golang extensions etc. I have been adamant about reporting to extensions that they should not activate themselves unless files they are interested in are opened. As of 2026 I would say most extensions I know of now work gracefully here.

Of course there is a very slight startup cost when opening VS Code since you have more extensions always on. But I'd say it is a very small hit. It's just "worth it" than to mess around with the profiles implementation, which feels like an afterthought rather than a proper planned implementation.

**Comment by sanmai-NL** [comment]
@thernstig Interesting. But for generalist developers using a very wide spectrum of extensions, juist the UI clutter of having all extensions enabled is major issue. Perhaps this also suggest the proper scope for this feature: allow UI layout profiles, no so much extension profiles. 

Another alternative I consider is to generate VS Code configuration using a more powerful language such as KCL. That would enable effective inheritance and platform-dependent logic.

**Comment by thernstig** [comment]
@sanmai-NL yes, it can slightly be problematic. But I disable some UI-clutter via right-clicking the Activity Bar or Status bar, and disabling many extensions I feel add no value there.

The general idea about profiles is good, but I also found it becomes problematic if you are in various repos that require various extensions. Some monorepos use many languages, so in those cases you would use lots of extensions. I feel disabling/enabling extensions (via profiles) thus make the UI "move around" in what icons etc. are seen. I find that more disturbing as a generalist, than to have many shown at once (but at stable locations) with some hidden.

**Comment by loderunner** [comment]
### Prettier/ESLint/TS6 vs. Oxfmt/Oxlint/TS7 use case

Many JS/TS projects use:
* Prettier
* ESLint
* TypeScript <7.0.0

The "cutting edge" toolchain is becoming:
* Oxfmt
* Oxlint
* TypeScript >=7.0.0

These tools have the same goal, but a completely different set of extensions and settings (most importantly per-language `editor.defaultFormatter`). I currently work on many JS/TS projects, with varying toolchains.

My current workflow involves manually defining per language formatters and linters in `${workspaceFolder}/.vscode/settings.json` for each project. E.g.

```json
{
  "editor.codeActionsOnSave": {
    "source.fixAll.oxc": "always"
  },
  "[javascript]": {
    "editor.defaultFormatter": "oxc.oxc-vscode"
  },
  "[typescript]": {
    "editor.defaultFormatter": "oxc.oxc-vscode"
  },
  "[json]": {
    "editor.defaultFormatter": "oxc.oxc-vscode"
  },
  "[jsonc]": {
    "editor.defaultFormatter": "oxc.oxc-vscode"
  },
  "[yaml]": {
    "editor.defaultFormatter": "oxc.oxc-vscode"
  },
  "[markdown]": {
    "editor.defaultFormatter": "oxc.oxc-vscode"
  }
}
```

I would like to have a Prettier/ESLint profile, and another Oxfmt/Oxlint profile, so that I can assign the profile to workspaces, and have both profiles inherit all of the other global shared settings from user `settings.json`.
